state-sync/aptos-data-client/src/aptosnet/mod.rs

```diff
@@ -146,46 +156,65 @@ impl AptosNetDataClient {
+        // Fetch and log the number of in-flight polls
+        let num_in_flight_polls = self.peer_states.read().num_in_flight_priority_polls();
+        update_in_flight_metrics(PRIORITIZED_PEER, num_in_flight_polls);
 
-        // Fetch the last polled high-priority peer
-        if let Some(peer) = self.peer_states.write().oldest_polled_priority_peer() {
-            peers_to_poll.push(peer);
+        // Ensure we don't go over the maximum number of in-flight priority polls
+        if num_in_flight_polls >= self.data_client_config.max_num_in_flight_priority_polls {
+            return Ok(None);
         }
 
-        // Fetch all new peers (i.e., those not yet polled)
-        for peer in self.get_all_connected_peers()? {
-            if !self.peer_states.read().already_polled_peer(&peer) {
-                peers_to_poll.push(peer);
-            }
-        }
+        // Get all connected prioritized peers
+        let mut priority_peers = self.get_all_connected_peers()?;
+        priority_peers.retain(|peer| self.peer_states.read().is_priority_peer(peer));
 
-        // Handle regular peer polling
-        if peers_to_poll.is_empty() {
-            // Always try and poll at least one peer
-            if let Some(peer) = self.peer_states.write().oldest_polled_regular_peer() {
-                peers_to_poll.push(peer);
-            }
-        } else {
-            // Poll regular peers at a 1/3 reduced frequency
-            sample!(SampleRate::Frequency(3), {
-                if let Some(peer) = self.peer_states.write().oldest_polled_regular_peer() {
-                    peers_to_poll.push(peer);
-                }
-            });
+        self.select_peer_to_poll(priority_peers)
+    }
+
+    /// Fetches the next regular peer to poll. The peer will be marked
+    /// as having an in-flight request.
+    fn fetch_regular_peer_to_poll(&self) -> Result<Option<PeerNetworkId>, Error> {
+        // Fetch the number of in-flight polls
+        let num_in_flight_polls = self.peer_states.read().num_in_flight_regular_polls();
+        update_in_flight_metrics(REGULAR_PEER, num_in_flight_polls);
+
+        // Ensure we don't go over the maximum number of in-flight regular polls
+        if num_in_flight_polls >= self.data_client_config.max_num_in_flight_regular_polls {
+            return Ok(None);
         }
 
-        // Mark all peers as polled
-        for peer in &peers_to_poll {
-            self.peer_states.write().mark_peer_as_polled(peer);
+        // Get all connected regular peers
+        let mut regular_peers = self.get_all_connected_peers()?;
+        regular_peers.retain(|peer| !self.peer_states.read().is_priority_peer(peer));
+
+        self.select_peer_to_poll(regular_peers)
+    }
+
+    /// Randomly selects a peer to poll that does not have an in-flight request
+    fn select_peer_to_poll(
+        &self,
+        mut peers: Vec<PeerNetworkId>,
+    ) -> Result<Option<PeerNetworkId>, Error> {
```

Comment:
IIUC, previously we used to poll atleast 2 peers per poll, now we are doing just 1. Is that a concern around increasing state sync latency?

Reply:
Also, does it makes sense to have a configurable value around number of peers we want to poll ?

Reply:
> IIUC, previously we used to poll atleast 2 peers per poll, now we are doing just 1. Is that a concern around increasing state sync latency?

IIRC (🤔) the previous world only polled 1 peer per round on average. The sampling we have now makes it 1.3 peers (i.e., 3 priority and 1 regular peers on average per 3 poll calls). This should be enough.

> Also, does it makes sense to have a configurable value around number of peers we want to poll ?

Technically, we already do. We just increase the poll frequency (which is configurable) 😄. I'm not a big of fan adding another configuration variable. Seems like overkill.